### DISCUSSION TITLE
Sources of Performance Improvement

### DISCUSSION THREAD

**Discussion by gyh1997127** [verificarlo/verificarlo#283] [Q&A] [answered]
Dear devs,

I read the paper "Verificarlo: checking floating point accuracy through Monte Carlo Arithmetic", and I have some questions.

The paper reported that Verificarlo had a ~116x overhead over IEEE.
![Screen Shot 2021-09-27 at 11 46 35 AM](https://user-images.githubusercontent.com/41491404/134967157-21f96570-ef0c-49ab-9365-4d1685994c53.png)

Out of curiosity, I ran the test myself using the latest verificarlo, the code is:
```
void compensated_sum() {
  int test_num = 1000000;
  float f[test_num];
  float sum = f[0];
  for (int i = 0; i < test_num; i++)
    f[i] = rand()/(float)(RAND_MAX)*(24.f + 1.f);
  float c = 0.0, y, t;

  for (int j = 0; j < 128; j++) {
    for (int i = 1; i < test_num; i++) {
      y = f[i] - c;
      t = sum + y;
      c = (t - sum) - y;
      sum = t;
    }
  }
}
```
Compiled with `verificarlo-c++ test.cpp -Wdouble-promotion -O0`
Ran with: 
- time VFC_BACKENDS="libinterflop_mca.so" ./a.out
- time VFC_BACKENDS="libinterflop_ieee.so" ./a.out
on Intel(R) Xeon(R) Silver 4214 CPU @ 2.20GHz

The results are a bit shocking:
- MCA reported sys time of 19.475s, 0.152s/sample
- IEEE reported sys time of 3s, 0.023s/sample
- only a 6.6x overhead

My questions are:
1. Since I ran the program without changing the virtual precision, can the dramatic performance improvement be because that the program is run in IEEE double? 
2. What else may be contributing to the significant performance improvement?
3. How do I vary the virtual precision during runtime?

Thank you for your time!

**Comment by titeup** [comment, marked as answer]
Hello, 

1- In the 2015 paper this was our first result using the mpfr backend, since then many progress has been made using quad format and finely tuned arithmetic to add the noise and rounding. However we haven't check the performance for a long time, so I cannot tell how accurate is your measurement without testing myself
2-I think I already told most of it in 1/. The code is reasonably  well documented and easy to read, you can check in https://github.com/verificarlo/verificarlo/blob/7fab3ff93ab4596c456142ae270cd4fa2acf1137/src/backends/interflop-mca/interflop_mca.c
3-We reorganize the documentation recently and I realize there is no short way to reach the info, here it is: https://github.com/verificarlo/verificarlo/blob/master/doc/02-Backends.md#mca-backend-libinterflop_mcaso

Additionally if you want to learn and play around the best way is to look at the tutorial: https://github.com/verificarlo/verificarlo/wiki/Tutorials

**Comment by gyh1997127** [comment]
Thank you for the response. 

I'm still a bit confused regarding questions 3. I see that the --precision-binary32 and --precision-binary64 options allow one to configure the precision before execution. Is there a way to modify the precisions during execution? Perhaps by invoking a function such as _set_precision(int i) of some sort?

**Comment by titeup** [comment]
Yes indeed you can, this is what we have done in vprec paper:
Chatelain, Yohan & Petit, Eric & de Oliveira Castro, Pablo & Lartigue, Ghislain & Defour, David. (2019). Automatic Exploration of Reduced Floating-Point Representations in Iterative Methods. 10.1007/978-3-030-29400-7_34.

**Comment by titeup** [comment]
Since it appears to not be an issue, I will convert this thread to a discussion.
Thanks,
E-
